jepsen: table creation fails with transaction retry error

The following tests appear to have failed:

[#530931](https://teamcity.cockroachdb.com/viewLog.html?buildId=530931):


```
--- FAIL: Jepsen/Jepsen: JepsenBank-multitable: JepsenBank-multitable/subcritical-skews+start-kill-2 (11.845s)
None
--- FAIL: Jepsen/Jepsen: JepsenG2: JepsenG2/split (14.752s)
None
--- FAIL: Jepsen/Jepsen: JepsenSequential: JepsenSequential/strobe-skews (62.548s)
None

```

Please assign, take a look and update the issue accordingly.

--- a/pkg/internal/client/txn.go
+++ b/pkg/internal/client/txn.go
@@ -1212,10 +1212,34 @@ func (txn *Txn) GenerateForcedRetryableError(msg string) error {
 		txn.ID(),
 		roachpb.MakeTransaction(
 			txn.DebugName(),
-			nil, // baseKey
+			nil, // aseKey
 			txn.UserPriority(),
 			txn.Isolation(),
 			txn.db.clock.Now(),
 			txn.db.clock.MaxOffset().Nanoseconds(),
 		))
+}
+
+// IsSerializablePushAndRefreshNotPossible returns true if the transaction is
+// serializable, its timestamp has been pushed and there's no chance that
+// refreshing the read spans will succeed later (thus allowing the transaction
+// to commit and not be restarted). Used to detect whether the txn is guaranteed
+// to get a retriable error later.
+//
+// Note that this method allows for false negatives: sometimes the client only
+// figures out that it's been pushed when it sends an EndTransaction - i.e. it's
+// possible for the txn to have been pushed asynchoronously by some other
+// operation (usually, but not exclusively, by a high-priority txn with
+// conflicting writes).
+func (txn *Txn) IsSerializablePushAndRefreshNotPossible() bool {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+
+	origTimestamp := txn.Proto().OrigTimestamp
+	origTimestamp.Forward(txn.Proto().RefreshedTimestamp)
+	isTxnPushed := txn.Proto().Timestamp != origTimestamp
+	// We check OrigTimestampWasObserved here because, if that's set, refreshing
+	// of reads is not performed.
+	return txn.Proto().Isolation == enginepb.SERIALIZABLE &&
+		isTxnPushed && txn.mu.Proto.OrigTimestampWasObserved
 }

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -372,6 +372,29 @@ func (ex *connExecutor) execStmtInOpenState(
 		if err := res.Err(); err != nil {
 			return makeErrEvent(err)
 		}
+
+		txn := ex.state.mu.txn
+		if !os.ImplicitTxn.Get() && txn.IsSerializablePushAndRefreshNotPossible() {
+			rc, canAutoRetry := ex.getRewindTxnCapability()
+			if canAutoRetry {
+				ev := eventRetriableErr{
+					IsCommit:     fsm.FromBool(isCommit(stmt.AST)),
+					CanAutoRetry: fsm.FromBool(canAutoRetry),
+				}
+				txn.Proto().Restart(0 /* userPriority */, 0 /* upgradePriority */, ex.server.cfg.Clock.Now())
+				payload := eventRetriableErrPayload{
+					err: roachpb.NewHandledRetryableTxnError(
+						"serializable transaction timestamp pushed (detected by connExecutor)",
+						txn.ID(),
+						// No updated transaction required; we've already manually updated our
+						// client.Txn.
+						roachpb.Transaction{},
+					),
+					rewCap: rc,
+				}
+				return ev, payload, nil
+			}
+		}
 	}
 
 	// No event was generated.

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -801,25 +801,6 @@ func (t *logicTest) outf(format string, args ...interface{}) {
 // It returns a cleanup function to be run when the credentials
 // are no longer needed.
 func (t *logicTest) setUser(user string) func() {
-	var outDBName string
-
-	if t.db != nil {
-		var inDBName string
-
-		if err := t.db.QueryRow("SHOW DATABASE").Scan(&inDBName); err != nil {
-			t.Fatal(err)
-		}
-
-		defer func() {
-			if inDBName != outDBName {
-				// Propagate the DATABASE setting to the newly-live connection.
-				if _, err := t.db.Exec(fmt.Sprintf("SET DATABASE = '%s'", inDBName)); err != nil {
-					t.Fatal(err)
-				}
-			}
-		}()
-	}
-
 	if t.clients == nil {
 		t.clients = map[string]*gosql.DB{}
 	}
@@ -827,17 +808,16 @@ func (t *logicTest) setUser(user string) func() {
 		t.db = db
 		t.user = user
 
-		if err := t.db.QueryRow("SHOW DATABASE").Scan(&outDBName); err != nil {
-			t.Fatal(err)
-		}
-
 		// No cleanup necessary, but return a no-op func to avoid nil pointer dereference.
 		return func() {}
 	}
 
 	addr := t.cluster.Server(t.nodeIdx).ServingAddr()
 	pgURL, cleanupFunc := sqlutils.PGUrl(t.t, addr, "TestLogic", url.User(user))
-	db, err := gosql.Open("postgres", pgURL.String())
+	// Connect to the test database by default, to match the default test
+	// connection (which creates and uses that database).
+	urlStr := pgURL.String() + "&database=test"
+	db, err := gosql.Open("postgres", urlStr)
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/pkg/sql/logictest/testdata/logic_test/alter_table
+++ b/pkg/sql/logictest/testdata/logic_test/alter_table
@@ -594,6 +594,11 @@ INSERT INTO privs VALUES (1)
 
 user testuser
 
+query T
+SHOW DATABASE
+----
+test
+
 statement error user testuser does not have CREATE privilege on relation privs
 ALTER TABLE privs ADD c INT
 

--- a/pkg/sql/logictest/testdata/logic_test/pgoidtype
+++ b/pkg/sql/logictest/testdata/logic_test/pgoidtype
@@ -221,17 +221,20 @@ user root
 statement ok
 CREATE DATABASE otherdb
 
-statement ok
-SET DATABASE = otherdb
-
 ## a non-root user can't get the OID of a table from a different database
 
 user testuser
+
+statement ok
+SET DATABASE = otherdb
 
 query error pgcode 42P01 relation "a" does not exist
 SELECT 'a'::regclass
 
 user root
+
+statement ok
+SET DATABASE = otherdb
 
 statement ok
 CREATE TABLE a (id INT PRIMARY KEY, foo STRING)

--- a/pkg/sql/logictest/testdata/logic_test/txn
+++ b/pkg/sql/logictest/testdata/logic_test/txn
@@ -494,6 +494,10 @@ snapshot
 statement ok
 COMMIT
 
+# restore default
+statement ok
+SET DEFAULT_TRANSACTION_ISOLATION TO 'SERIALIZABLE'
+
 # SHOW TRANSACTION STATUS
 
 query T
@@ -977,3 +981,7 @@ SELECT CRDB_INTERNAL.FORCE_RETRY('1s':::INTERVAL)
 
 statement ok
 ROLLBACK
+
+# restore the default
+statement ok
+SET default_transaction_read_only = false

--- a/pkg/sql/logictest/testdata/logic_test/txn_retry
+++ b/pkg/sql/logictest/testdata/logic_test/txn_retry
@@ -1,0 +1,37 @@
+# LogicTest: default
+
+# Check that we auto-retry pushed transactions which can't be refreshed - if
+# they're pushed while we can still auto-retry them.
+subtest autoretry-on-push-first-batch
+
+statement ok
+CREATE TABLE test_retry (
+  k INT PRIMARY KEY
+)
+
+statement ok
+GRANT ALL ON test_retry TO testuser
+
+# Start a txn (and fix a timestamp).
+statement ok
+BEGIN
+
+# On a different connection, do a read at a higher timestamp.
+#
+# We use dont_inherit_database, as otherwise we'd run some statements under the
+# hood that break the auto-retry that we'll need later.
+user testuser dont_inherit_database
+
+statement ok
+SELECT * FROM test.test_retry
+
+user root dont_inherit_database
+
+# Run a cluster_logical_timestamp(), so that the transaction "observes its
+# commit timestamp" and so can't be refreshed, and the do an insert that will
+# cause the txn to be pushed.
+statement ok
+SELECT cluster_logical_timestamp(); INSERT INTO test_retry VALUES (1);
+
+statement ok
+COMMIT

